// hexo-custom-autolinker/lib/filter.js
var Autolinker = require('autolinker');
var cheerio = require('cheerio');
var _ = require('lodash');
var urlparser = require('url');
var ResponsiveVideo = require('responsive-video');

var _defaultConfig = {
    replaceFn : function( autolinker, match ) {
        
        if (match.getType() == 'url') {
            var type = 'url';
            //console.log(match)
            var shortened_text = match.matchedText;
            if (match.matchedText.length > 30) {
              shortened_text = match.matchedText.slice(0,28) + '...';
            }
             
            if (match.getUrl().indexOf('twitter.com') != -1) {
              var parsed = urlparser.parse(match.getUrl());
              var tweet_id = parsed['path'].split('/')[3];
              var type = 'url';
              
              return '<div class="twitter_wrap tweet-link" data-tweet-id="'+tweet_id+'"><a rel="nofollow" href="'+match.url+'" class="tweet_url">'+shortened_text+'</a></div>';
            }
            if ((match.getUrl().indexOf('facebook.com') != -1) && ((match.getUrl().indexOf('posts') != -1)) || (match.getUrl().indexOf('photos') != -1) || (match.getUrl().indexOf('events') != -1) || (match.getUrl().indexOf('videos') != -1)) {
              return '<div class="fb_post_wrap"><div class="fb-post" data-href="'+match.url+'" data-width="auto" data-show-text="true"></div><a rel="nofollow" href="'+match.url+'" class="fb_url">'+shortened_text+'</a></div>';

            }
            if ((match.getUrl().indexOf('youtube.com') != -1) || (match.getUrl().indexOf('youtu.be') != -1) || (match.getUrl().indexOf('vimeo.com') != -1) || (match.getUrl().indexOf('vimeo.com') != -1) || (match.getUrl().indexOf('dai.ly') != -1)) {
                var parsed = urlparser.parse(match.getUrl());
                
                if (match.getUrl().indexOf('vimeo.com') != -1) {
                    var type = 'vimeo';
                    var id = parsed['path'].replace('/', '');
                }
                if (match.getUrl().indexOf('dai.ly') != -1) {
                    var type = 'dailymotion';
                    var id = parsed['path'].replace('/', '');
                }

                if ((match.getUrl().indexOf('youtube.com') != -1) || (match.getUrl().indexOf('youtu.be') != -1)) {
                    var type = 'youtube';
                    if (match.getUrl().indexOf('youtu.be') != -1) {
                        var id = parsed['path'].replace('/', '');
                    }
                    else {
                        var id = parsed['query'].replace('v=', '');
                    }
                }
                return ResponsiveVideo(id, type);
            }
        }
    },
    newWindow: true,
    stripPrefix: false,
    phone: false,
    twitter: false,
    hashtag: false
};

module.exports = function(source) {
    var hexo = this;
    
    this.config = this.config || {};

    var config = this.config.autolinker || {};

    config = _.extend({}, _defaultConfig, config);

    // The following code could be separated out into another module, but that seems like overkill for now.
    var $ = cheerio.load(source);
    $("img").each(function() {
      if (!$(this).hasClass('lazy-ignore')) {
       
        if ($(this).attr("src")) {
          $(this).attr("data-src", $(this).attr("src"));
          $(this).attr("src", "");
          $(this).attr('class', 'lazy');
        }
      }
    });
    
    
    source = $.html();
    source = Autolinker.link(source, config);
    
    return source;
};
